### PR TITLE
[QGL, JMENanoAOD] PUPPI-compatible QGTagger producer

### PR DESCRIPTION
**Pull request by nurfikri89**
#### PR description:

The main goal of this PR is to modify QGTagger producer (used to compute Quark-Gluon (QG) Likelihood) so that it is compatible with PUPPI jets. The `QGTagger` calculates three input variables, used to compute the likelihood, from PF constituents of jets. The changes in this PR allows the QGTagger to compute the input variables with PUPPI weights applied on the PF constituents by accessing the weights from the provided `ValueMap<float>`. This can also be done for any constituents weights, not just PUPPI.  Additionally, this PR makes it optional to not compute the likelihood and calculate just the three input variables for users who wants only them. 

With these modifications of the QGTagger, it should not affect CHS jets so we do not expect to see any change to the output of workflows. The QGTagger is currently scheduled to run only for AK4 CHS jets at the MiniAOD and (Run-2) NanoAOD production level.

Other changes:
- some fixes for JMENanoAOD. 
- fine-tune histograms in nanoDQM for JMENano.

#### PR validation:
- Validation plots comparing the input variables before and after the fix can be found in this JMAR meeting [contribution](https://indico.cern.ch/event/1245910/#7-virtual-quark-gluon-likeliho). 
- passes the usual runTheMatrix test: `runTheMatrix.py -l limited -i all --ibeos`
- passes reMiniAOD and reNanoAOD workflows: `runTheMatrix.py -i all --ibeos -l 1325.518,2500.312`
- passes the JMENano workflows: `runTheMatrix.py -i all --ibeos -l 10224.15,11024.15,25202.15,11634.15`

### CODE REVIEW

**Comment by swertz** [comment]
enable nano

**Comment by swertz** [comment]
please test

**Comment by swertz** [comment]
please test

File: PhysicsTools/NanoAOD/python/custom_jme_cff.py [hunk 1]
```diff
@@ -185,7 +185,7 @@
   particleNetAK4_CvsL = Var("?(pt>=15)?bDiscriminator('pfParticleNetAK4DiscriminatorsJetTags:CvsL'):-1",float,doc="ParticleNetAK4 tagger c vs udsg discriminator",precision=10),
   particleNetAK4_CvsB = Var("?(pt>=15)?bDiscriminator('pfParticleNetAK4DiscriminatorsJetTags:CvsB'):-1",float,doc="ParticleNetAK4 tagger c vs b discriminator",precision=10),
   particleNetAK4_QvsG = Var("?(pt>=15)?bDiscriminator('pfParticleNetAK4DiscriminatorsJetTags:QvsG'):-1",float,doc="ParticleNetAK4 tagger uds vs g discriminator",precision=10),
-  particleNetAK4_G = Var("?(pt>=15)?bDiscriminator('pfParticleNetAK4DiscriminatorsJetTags:probg'):-1",float,doc="ParticleNetAK4 tagger g raw score",precision=10),
+  particleNetAK4_G = Var("?(pt>=15)?bDiscriminator('pfParticleNetAK4JetTags:probg'):-1",float,doc="ParticleNetAK4 tagger g raw score",precision=10),
```
Comment (swertz) [inline PhysicsTools/NanoAOD/python/custom_jme_cff.py:188]: The changes to the b-tagging and particleNet taggers seem unrelated to the QGL, are they expected?

File: RecoJets/JetProducers/python/QGTagger_cfi.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
 import FWCore.ParameterSet.Config as cms
 
 QGTagger = cms.EDProducer('QGTagger',
-  srcJets		= cms.InputTag('ak4PFJetsCHS'),
```
Comment (swertz) [inline RecoJets/JetProducers/python/QGTagger_cfi.py:4]: In principle with the `fillDescriptions` (and with default values), the _cfi could be removed, no?

**Review by swertz — COMMENTED** [review]
Thanks @nurfikri89 , changes seem fine and as expected there is no difference in Nano workflows; I only have two small comments.

File: PhysicsTools/NanoAOD/python/custom_jme_cff.py (hunk 1 above)
Comment (nurfikri89) [reply PhysicsTools/NanoAOD/python/custom_jme_cff.py:188]: They're unrelated yes. I just took this opportunity to fix this part since I'm modifying the `custom_jme_cff.py` script to propagate the changes made to `QGTagger`.

File: RecoJets/JetProducers/python/QGTagger_cfi.py (hunk 2 above)
Comment (nurfikri89) [reply RecoJets/JetProducers/python/QGTagger_cfi.py:4]: I assume so yes. This `_cfi` has always been there for a long time but maybe because someone used it to define the `QGTaggerTask`.

Comment (swertz) [reply RecoJets/JetProducers/python/QGTagger_cfi.py:4]: The only use I find is this: https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L244

Perhaps we can take this opportunity to replace it with an import+clone and define the task there?

Comment (nurfikri89) [reply RecoJets/JetProducers/python/QGTagger_cfi.py:4]: Yeah. Okay I can do it.

Comment (swertz) [reply RecoJets/JetProducers/python/QGTagger_cfi.py:4]: Thanks!

**Comment by swertz** [comment]
type jetmet

**Comment by swertz** [comment]
+1

No differences seen, as expected.

**Comment by mandrenguyen** [comment]
+1
xpog business

**Comment by perrotta** [comment]
+1
